File: area_scan_3d_camera/Basic/CapturePointCloudHDR/CapturePointCloudHDR.cpp
/*******************************************************************************
 *BSD 3-Clause License
 *
 *Copyright (c) 2016-2024, Mech-Mind Robotics
 *All rights reserved.
 *
 *Redistribution and use in source and binary forms, with or without
 *modification, are permitted provided that the following conditions are met:
 *
 *1. Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *2. Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 *3. Neither the name of the copyright holder nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 *THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 *AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 *FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 *DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 *SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 *OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 *OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************************************************************/

/*
With this sample, you can set multiple exposure times, and then obtain and save the point cloud.
*/

#include "area_scan_3d_camera/parameters/Scanning3D.h"
#include "area_scan_3d_camera/api_util.h"
#include "area_scan_3d_camera/Camera.h"

int main()
{
    mmind::eye::Camera camera;
    if (!findAndConnect(camera))
        return -1;

    // Set the 3D exposure times
    mmind::eye::UserSet& currentUserSet = camera.currentUserSet();
    showError(currentUserSet.setFloatArrayValue(
        mmind::eye::scanning3d_setting::ExposureSequence::name, std::vector<double>{5, 10}));

    if (!confirmCapture3D()) {
        camera.disconnect();
        return 0;
    }
    mmind::eye::Frame3D frame3D;
    showError(camera.capture3D(frame3D));

    const std::string pointCloudFile = "PointCloud.ply";
    showError(frame3D.saveUntexturedPointCloud(mmind::eye::FileFormat::PLY, pointCloudFile));
    std::cout << "Capture and save the point cloud: " << pointCloudFile << std::endl;

    camera.disconnect();
    std::cout << "Disconnected from the camera successfully." << std::endl;

    return 0;
}

File: area_scan_3d_camera/basic/capture_point_cloud_hdr.py
# With this sample, you can set multiple exposure times, and then obtain and save the point cloud.

from mecheye.shared import *
from mecheye.area_scan_3d_camera import *
from mecheye.area_scan_3d_camera_utils import find_and_connect, confirm_capture_3d


class CapturePointCloudHDR(object):
    def __init__(self):
        self.camera = Camera()
        self.frame_2d_and_3d = Frame2DAnd3D()

    def capture_point_cloud(self):
        point_cloud_file = "PointCloud.ply"
        success_message = "Capture and save the untextured point cloud to {}.".format(point_cloud_file)
        show_error(
            self.frame_2d_and_3d.frame_3d().save_untextured_point_cloud(FileFormat_PLY, point_cloud_file), success_message)

    def capture_textured_point_cloud(self):
        textured_point_cloud_file = "TexturedPointCloud.ply"
        success_message = "Capture and save the textured point cloud to {}".format(textured_point_cloud_file)
        show_error(self.frame_2d_and_3d.save_textured_point_cloud(FileFormat_PLY,
                                                                  textured_point_cloud_file),success_message)
        
    def capture_point_cloud_hdr(self):
        # Set 3D Exposure Sequence.
        current_user_set = self.camera.current_user_set()
        error = current_user_set.set_float_array_value(
            Scanning3DExposureSequence.name, [5, 10])
        show_error(error)
        self.capture_point_cloud()
        self.capture_textured_point_cloud()

    def main(self):
        if find_and_connect(self.camera):
            if not confirm_capture_3d():
                return
            show_error(self.camera.capture_2d_and_3d(self.frame_2d_and_3d))
            self.capture_point_cloud_hdr()
            print("Disconnected from the camera successfully.")


if __name__ == '__main__':
    a = CapturePointCloudHDR()
    a.main()
